### PR TITLE
4676 add translation files to service provider

### LINKED ISSUE #4676: Relocate embed to lower dependency
Add rust embed for locales in service layer so that it is accessible.

Part of epic #4682

### PR DESCRIPTION
**Pull request by fergie-nz**
Fixes #4676 

# 👩🏻‍💻 What does this PR do?

Adds localisations to service layer in back end.

Still some work to do and this is as yet incomplete:

- follow up work will continue as part of epic #4682 

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] _(e.g.)_ Central Sync server with 1 Legacy Desktop remote site and 1 OMS remote site running this PR
- [ ] _(e.g.)_ This sample datafile: _google drive link_
- [ ] _(e.g.)_ Open a requisition with some lines
- [ ] _(e.g.)_ Make a couple invoices supplying some amount of those lines
- [ ] _(e.g.)_ Review that "issued" column is the sum of the amount already issued in invoices for this requisition

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: server/service/src/localisations.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    // Load translations from embedded files
+    pub fn load_translations(&mut self) -> Result<(), std::io::Error> {
+        // Languages - need to extract these from the files themselves?
+        let mut languages = Vec::new();
+
+        // read all dirs within locales dir
+        let folder_paths: Vec<String> = fs::read_dir("../client/packages/host/dist/locales")
```
Comment (andreievg) [inline server/service/src/localisations.rs:38]: This is not available in production, I think you would use `EmbeddedLocalisations::iter()`

File: server/service/src/localisations.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            .unwrap()
+            .filter_map(|e| e.ok())
+            .map(|e: fs::DirEntry| e.path().as_os_str().to_string_lossy().to_string().split('/').last().unwrap().to_string())
```
Comment (andreievg) [inline server/service/src/localisations.rs:41]: using `::iter()` would return `Cow` str, so should be easier to work with !

File: server/service/src/localisations.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        for lang in languages {
+            if let Some(content) = EmbeddedLocalisations::get(&format!("locales/{}/common.json", &lang)) {
+                let json_data = content.data;
+                let translations: HashMap<String, String> = serde_json::from_slice(&json_data).unwrap();
+                self.translations.insert(lang.to_string(), translations);
+
+            }
+        }
+        // later need to think about how to concatonate all translation json files per language
```
Comment (andreievg) [inline server/service/src/localisations.rs:56]: I think we talked about this, we still need to access each string by 'name space' so shouldn't concatenate.

File: server/service/src/localisations.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+        for path in folder_paths {
+            languages.push(path)
+        }
+
+        for lang in languages {
+            if let Some(content) = EmbeddedLocalisations::get(&format!("locales/{}/common.json", &lang)) {
+                let json_data = content.data;
+                let translations: HashMap<String, String> = serde_json::from_slice(&json_data).unwrap();
+                self.translations.insert(lang.to_string(), translations);
```
Comment (andreievg) [inline server/service/src/localisations.rs:52]: this will replace previous names space

File: server/service/src/localisations.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+            }
+        }
+        // later need to think about how to concatonate all translation json files per language
+
+        Ok(())
+    }
+
+    // Get a translation for a given key and language
+    pub fn get_translation(&self, key: &str, language: &str) -> String {
```
Comment (andreievg) [inline server/service/src/localisations.rs:62]: I think there would be 4 parameters:

language, namespace, key, fallback

**Review by andreievg — CHANGES_REQUESTED** [review]
Thanks in general look good, but a few requests.

**Comment by andreievg** [comment]
I was just playing around a little bit, and had something like this (for reference, I thought it would help with readability), untested:

```rust
type Translations = HashMap<
    /* Language */ String,
    HashMap<
        /* Namespace */ String,
        HashMap</* TranslationKey */ String, /* TranslationString */ String>,
    >,
>;

const DEFAULT_LANGUAGE: &str = "en";

fn example(language: &str, namespace: &str, key: &str, placeholder: &str) -> String {
    let translation: Translations = HashMap::new();

    translation
        .get(language)
        .map(|l| l.get(namespace).map(|n| n.get(key)))
        .flatten()
        .flatten()
        .map(String::to_string)
        .unwrap_or_else(|| {
            if DEFAULT_LANGUAGE == language {
                placeholder.to_string()
            } else {
                example(language, namespace, key, placeholder)
            }
        })
        .clone()
}
```

File: server/service/src/localisations.rs (hunk 4 above)
Comment (fergie-nz) [reply server/service/src/localisations.rs]: Nice, I see now. I've updated to have a more nested hashmap so we can access via language, namespace, and key

File: server/service/src/service_provider.rs [hunk 6]
```diff
@@ -180,9 +139,9 @@ impl ServiceProvider {
         sync_trigger: SyncTrigger,
         site_is_initialised_trigger: SiteIsInitialisedTrigger,
     ) -> Self {
-        if let Some(content) = Localisations::get("locales/en/common.json") {
-            println!("{:?}", content.data);
-        }
+        let mut localisations = Localisations::new();
```
Comment (andreievg) [inline server/service/src/service_provider.rs:142]: load_translations could be with ::new() -> inline when boxing

**Review by andreievg — APPROVED** [review]
Looks great, thanks for changes !

**Pull request merged**